### PR TITLE
Try to Remove autogenerate requirements files.

### PR DESCRIPTION
**Pull request by Carreau**
Where possible in the github action use `.[...]` syntax

With the exeption  of lint that was not installing networkx and now does I hope that this should be identical

It does sometime merge a few pip command into a single one.

Note that some of those files are not autogenerated, so leave them.

One of the limitation is that the instructions to intall with conda (`conda install -c conda-forge --file requirements/extra.txt`) don't make sens anymore.

This would replace #7483

### CODE REVIEW

File: CONTRIBUTING.rst [hunk 1]
```diff
@@ -44,16 +44,16 @@ Development Workflow
          python -m venv networkx-dev
          # Activate it
          source networkx-dev/bin/activate
-         # Install main development and runtime dependencies of networkx
-         pip install -r requirements/default.txt -r requirements/test.txt -r requirements/developer.txt
          #
+         #
+         # Build and install networkx from source
+         # this will also intall required dependencies.
```
Comment (Peiffap) [inline CONTRIBUTING.rst:50]:
```suggestion
         # this will also install required dependencies.
```

File: doc/README.rst [hunk 2]
```diff
@@ -9,17 +9,14 @@ for both the stable and the latest (i.e., development) releases.
 Instructions
 ~~~~~~~~~~~~
 
-After installing NetworkX and its dependencies, install the Python
-packages needed to build the documentation by entering the root
-directory and executing::
+To install NetworkX, its dependencies and the one necessary to build
```
Comment (Peiffap) [inline doc/README.rst:12]:
```suggestion
To install NetworkX, its dependencies and the ones necessary to build
```

File: doc/README.rst [hunk 3]
```diff
@@ -9,17 +9,14 @@ for both the stable and the latest (i.e., development) releases.
...
+the docs you can enter the root directory and execute::
 
-    pip install -r requirements/doc.txt
+    pip install .[doc]
 
-Building the example gallery additionally requires the dependencies
-listed in ``requirements/extra.txt`` and ``requirements/example.txt``::
+Building the example gallery additionally requires the ``extra`` and ``example``::
```
Comment (Peiffap) [inline doc/README.rst:17]:
```suggestion
Building the example gallery additionally requires the ``extra`` and ``example`` requirements::
```

**Review by Peiffap — COMMENTED** [review]
This looks much cleaner! I didn't check the actual content deeply yet but I caught some typos.

**Review by rossbar — COMMENTED** [review]
I'm +1 on figuring out how to remove the requirements file autogeneration, but I'm -1 on removing requirements files in general.

I personally find the workflow with requirements files (where tab complete works) much more convenient when setting up an environment than relying on the `extra_requires` keys.

**Comment by Carreau** [comment]
> I personally find the workflow with requirements files (where tab complete works) much more convenient when setting up an environment than relying on the `extra_requires` keys.

Well I think this is a pip/shell issue if you can't complete extras. 

`https://github.com/pypa/pip/pull/12761` may help (using backquotes to not crosslink unrelated issues).

**Comment by Carreau** [comment]
> What would the instructions for installing via conda/mamba look like?
> Should we keep the requirements files around just for that?

I think it can be to create the conda env and run pip inside. 

There are upstream issue to add support to pyproject.toml to `conda --file`: https://github.com/conda/conda/issues/10633 but there are problemwith mapping package names anyway wether using requirements.txt of any other non-conda native req file. 

If anything I think that having a environment.yml autogenerated would be ok.

**Comment by Carreau** [comment]
would #7495 that only update github workflow to not reference requirements.txrt be a good first step ?

**Pull request closed without merging**